src/data/markdown/docs/01 guides/05 Testing Guides/04 Large-scale k6 testing.md

```diff
@@ -0,0 +1,484 @@
+
+<div class="code-group">
+
+```text
+WARN[0013] Request Failed       error="Get http://test.k6.io: read tcp 172.31.72.209:35288->63.32.205.136:80: read: connection reset by peer"
+```
+
+</div>
+
+### context deadline exceeded
+
+Error like this happens when k6 was able to send a request, but the target system didn't respond in time. The default timeout in k6 is 60 seconds. If your system doesn't produce the response in this timeframe, this error will appear.
+
+<div class="code-group">
+
+```
+WARN[0064] Request Failed    error="Get http://test.k6.io: context deadline exceeded"
+```
+
+</div>
+
+### dial tcp 52.18.24.222:80: i/o timeout
+
+This is a similar error to the one above, but in this case k6 wasn't even able to make a request. The target system isn't able to establish a connection.
+
+
+<div class="code-group">
+
+```
+WARN[0057] Request Failed     error="Get http://pawel.staging.loadimpact.com/static/logo.svg?url=v3: dial tcp 52.18.24.222:80: i/o timeout"
+```
+
+</div>
+
+### socket: too many open files
+
+This error means that the load-generator machine isn't able to open TCP sockets because it reached the limit of open file descriptors.
+Make sure that your limit is set sufficiently high `ulimit -n 250000` should be enough for anyone :tm:
+
+<div class="code-group">
+
+```
+WARN[0034] Request Failed     error="Get http://99.81.83.131/static/logo.svg?ip=6: dial tcp 99.81.83.131:80: socket: too many open files"
+```
+
+</div>
+
+
+Note: you should decide what level of errors is acceptable. At large scale some errors are always present.
+If you make 50M requests with 100 failures, this is generally a good result (0.00002% errors).
+
+# Benchmarking k6 on AWS hardware
+
+We have executed few large tests on different EC2 machines to see how much load k6 can generate. 
+Our general observation is that k6 scales proportionally to the hardware. 2x larger machine is able to generate 2x more traffic.
+The limit to this scalability is in the number of open connections. Single Linux machine can open up to `65 535` sockets per IP. 
+This means that maximum of 65k requests can be executed simultaneously on a single machine.
+The RPS limit depends on the response time of the SUT. If responses are delivered in 100ms, the RPS limit is 650 000. 
+
+## Real-life test of a website.
+
+Testing the theoretical limits is fun, but that's not the point of this benchmark. 
+The point of this benchmark is to give users indication of how much traffic k6 can generate when executing complicated, real-life tests.
+For this purpose we have written a rather heavy [real-life website test](https://github.com/loadimpact/k6-hardware-benchmark/blob/master/scripts/website.js) that uses almost all k6 features.
+
+Setup:
+- All tests were executed on AWS EC2 instances
+- The "discardResponseBodies" recommendation was NOT used. (results would be better with this setting).
```

Comment:
I would argue you can try to do a test with this and see how much better they will be :). This way when someone doesn't use it  and looks at optimizing they can see some example of what the difference si

Reply:
yeah, that's a fair point, but I don't want to go back to testing for this article :-) 
I'll maintain this page, and will likely update it later.